charon/src/transform/normalize/transform_dyn_trait_calls.rs

```diff
@@ -0,0 +1,297 @@
+        let trait_pred = trait_ref.trait_decl_ref.clone().erase();
+        let vtable_ref = vtable_ref
+            .clone()
+            .substitute_with_self(&trait_pred.generics, &trait_ref.kind);
+        Ok(vtable_ref)
+    }
+
+    /// Get the correct field index for a method in the vtable struct.
+    fn get_method_field_id(
+        &self,
+        vtable_ref: &TypeDeclRef,
+        method_name: &TraitItemName,
+    ) -> Result<FieldId, Error> {
+        let vtable_name = vtable_ref.id.with_ctx(&self.ctx.into_fmt()).to_string();
+
+        let TypeId::Adt(vtable_id) = vtable_ref.id else {
+            raise_error!(
+                self.ctx,
+                self.span,
+                "Vtable reference {} is not an ADT type!",
+                vtable_name
+            );
+        };
+
+        // Get the vtable struct declaration by its ID
+        let Some(TypeDecl {
+            kind: TypeDeclKind::Struct(fields),
+            ..
+        }) = self.ctx.translated.type_decls.get(vtable_id)
+        else {
+            raise_error!(
+                self.ctx,
+                self.span,
+                "Definition of vtable struct {} is not found!",
+                vtable_name
+            );
+        };
+
+        // Find the index from the fields
+        for (index, field) in fields.iter().enumerate() {
+            if format!("method_{}", method_name) == *field.name.as_ref().unwrap() {
+                return Ok(FieldId::new(index));
+            }
+        }
+
+        // If we reach here, the method was not found in the vtable, which is an error
+        raise_error!(
+            self.ctx,
+            self.span,
+            "Could not determine method index for {} in vtable {}",
+            method_name,
+            vtable_name
+        );
+    }
+
+    /// Create local variables needed for vtable dispatch
+    /// TODO: use the vtable ref to create the correct types
+    fn create_vtable_locals(
+        &mut self,
+        vtable_ref: &TypeDeclRef,
+        method_ptr_ty: &Ty,
+    ) -> (Place, Place) {
+        // Create vtable type - for now use a raw pointer as placeholder
+        // In complete implementation this would be the actual vtable struct type
+        let vtable_ty = Ty::new(TyKind::RawPtr(
+            Ty::new(TyKind::Adt(vtable_ref.clone())),
+            RefKind::Shared,
+        ));
+
+        let vtable_place = self.locals.new_var(None, vtable_ty);
+        // push the storage-live statements for the new locals
+        self.statements.push(Statement {
+            span: self.span,
+            kind: StatementKind::StorageLive(vtable_place.as_local().unwrap()),
+            comments_before: vec![],
+        });
+
+        let method_ptr_place = self.locals.new_var(None, method_ptr_ty.clone());
+        self.statements.push(Statement {
+            span: self.span,
+            kind: StatementKind::StorageLive(method_ptr_place.as_local().unwrap()),
+            comments_before: vec![],
+        });
```

Comment:
The main point of `UllbcStatementTransformCtx` is to simplify this:
```suggestion
        let vtable_place = self.fresh_var(None, vtable_ty);
        let method_ptr_place = self.fresh_var(None, method_ptr_ty.clone());
```